packages/protocolbeat/src/panel-terminal/TerminalPanel.tsx

```diff
@@ -0,0 +1,102 @@
+import { useQuery } from '@tanstack/react-query'
+import { useEffect, useRef, useState } from 'react'
+import { useParams } from 'react-router-dom'
+import { executeCommand, getProject } from '../api/api'
+import { useTerminalStore } from '../store/store'
+
+export function TerminalPanel() {
+  const { project } = useParams()
+  const [discoveryChain, setDiscoveryChain] = useState<string | undefined>(
+    undefined,
+  )
+  const abortControllerRef = useRef<AbortController>()
+  const outputRef = useRef<HTMLDivElement>(null)
+  const { output, isRunning, addOutput, setIsRunning, clear } =
+    useTerminalStore()
+
+  if (!project) {
+    throw new Error('Cannot use component outside of project page!')
+  }
+
+  const getProjectResponse = useQuery({
+    queryKey: ['projects', project],
+    queryFn: () => getProject(project),
+  })
+  const chains = getProjectResponse.data?.chains ?? []
+
+  useEffect(() => {
+    clear()
+  }, [])
+
+  useEffect(() => {
+    return () => {
+      // Cleanup on unmount
+      abortControllerRef.current?.abort()
+    }
+  }, [])
+
+  useEffect(() => {
+    if (outputRef.current) {
+      outputRef.current.scrollTop = outputRef.current.scrollHeight
+    }
+  }, [output])
+
+  useEffect(() => {
+    if (chains[0] && !discoveryChain) {
+      setDiscoveryChain(chains[0].name)
+    }
+  }, [chains, discoveryChain])
+
+  async function handleExecute() {
+    if (!project || !discoveryChain || isRunning) return
+    clear()
+    setIsRunning(true)
+
+    try {
+      const stream = await executeCommand('discover', project, discoveryChain)
+      abortControllerRef.current = new AbortController()
+
+      const reader = stream.getReader()
+      while (!abortControllerRef.current.signal.aborted) {
+        const { value, done } = await reader.read()
+        if (done) break
```

Comment:
Shouldn't we somehow close the stream? Or pass the abortcontroller to fetch in executeCommand?

Reply:
This is no longer relevant since I switched to server-sent events.